### PR TITLE
Utilities: Advanced ocvalidate for error checking

### PR DESCRIPTION
**Pull request by PMheart**
I am sorry that I messed up my previous fork. Now I am doing it again.

Important Note: This is still a work in progress.

All checks performed can be found [here](https://github.com/PMheart/OpenCorePkg/blob/master/Utilities/ocvalidate/Checklist.md).

To all: Feel free to leave your comments here for any checks that I forgot. Thanks!

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -56,3 +56,4 @@ Legacy/BinDrivers/X64
 Legacy/BinDrivers/IA32
 Legacy/BootLoader/bin
 Legacy/BootSector/bin
+Utilities/ocvalidate/SampleCustom.plist
```
Comment (PMheart) [inline .gitignore:59]: This is added to prevent accidental commits of test materials. It should be removed before merging.

**Review by yezz123 — APPROVED** [review]

File: .gitignore (hunk 1 above)
Comment (vit9696) [reply .gitignore:59]: I do not mind this. Could be Utilities/ocvalidate/*.plist though.

Comment (PMheart) [reply .gitignore]: Done.

**Comment by junaedahmed** [comment]
Hello @PMheart ,

On ACPI > Add
Entry[N]->Path: .dsl filename suffix is not accepted.

only .aml suffix accepted would be more enforcing then .dsl suffix checking.

Thank you

**Comment by PMheart** [comment]
> Hello @PMheart ,
> 
> On ACPI > Add
> Entry[N]->Path: .dsl filename suffix is not accepted.
> 
> only .aml suffix accepted would be more enforcing then .dsl suffix checking.
> 
> Thank you

Hello,

Sorry, but no. In addition to `.aml`, we should also support other binary forms, including `.bin`. Only `.dsl` is not supported as it is plain text.

PMheart

**Comment by junaedahmed** [comment]
> > Hello @PMheart ,
> > On ACPI > Add
> > Entry[N]->Path: .dsl filename suffix is not accepted.
> > only .aml suffix accepted would be more enforcing then .dsl suffix checking.
> > Thank you
> 
> Hello,
> 
> Sorry, but no. In addition to `.aml`, we should also support other binary forms, including `.bin`. Only `.dsl` is not supported as it is plain text.
> 
> PMheart

Hi,
If I accidentally use .txt on a .dsl file will this produce any error. my intention of suggesting that was for that.

**Comment by PMheart** [comment]
> > > Hello @PMheart ,
> > > On ACPI > Add
> > > Entry[N]->Path: .dsl filename suffix is not accepted.
> > > only .aml suffix accepted would be more enforcing then .dsl suffix checking.
> > > Thank you
> > 
> > 
> > Hello,
> > Sorry, but no. In addition to `.aml`, we should also support other binary forms, including `.bin`. Only `.dsl` is not supported as it is plain text.
> > PMheart
> 
> Hi,
> If I accidentally use .txt on a .dsl file will this produce any error. my intention of suggesting that was for this.

Well, indeed it is a problem. But still, we should not pay quite much attention to this nuance, as firstly it’s not worth the time, secondly the special check for .dsl is to remind users to compile those dsl files from Docs/AcpiSamples to aml. That’s it.

PMheart

**Comment by PMheart** [comment]
Hopefully I have done the preliminary checks, with some `TODO` and `FIXME` to be handled after discussion. @vit9696, Please take a look. Thanks!

File: Utilities/ocvalidate/ValidateUEFI.c [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+  // Sanitise strings.
+  //
+  if (AsciiStrCmp (TextRenderer, "BuiltinGraphics") != 0
+    && AsciiStrCmp (TextRenderer, "BuiltinText") != 0
+    && AsciiStrCmp (TextRenderer, "SystemGraphics") != 0
+    && AsciiStrCmp (TextRenderer, "SystemText") != 0
+    && AsciiStrCmp (TextRenderer, "SystemGeneric") != 0) {
+    DEBUG ((DEBUG_WARN, "UEFI->Output->TextRenderer is illegal (Can only be BuiltinGraphics, BuiltinText, SystemGraphics, SystemText, or SystemGeneric)!\n"));
+    ++ErrorCount;
+  } else {
```
Comment (vit9696) [inline Utilities/ocvalidate/ValidateUEFI.c:97]: `else if`?

File: Utilities/ocvalidate/ValidateUEFI.c [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    //
+    // Check whether TextRenderer has System prefix.
+    //
+    if (AsciiStrnCmp (TextRenderer, "System", L_STR_LEN ("System")) == 0) {
+      IsTextRendererSystem = TRUE;
+    }
+  }
+
+  if (UserUefi->Apfs.EnableJumpstart
+    && (UserMisc->Security.ScanPolicy != 0 && (UserMisc->Security.ScanPolicy & OC_SCAN_ALLOW_FS_APFS) == 0)) {
```
Comment (vit9696) [inline Utilities/ocvalidate/ValidateUEFI.c:107]: This check is not correct. First part should be `(UserMisc->Security.ScanPolicy & OC_SCAN_FILE_SYSTEM_LOCK) != 0`. I.e. the condition should be "FS restrictions enabled and APFS FS disabled"

File: Utilities/ocvalidate/ValidateUEFI.c [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    DEBUG ((DEBUG_WARN, "UEFI->APFS->EnableJumpstart is enabled, but Misc->Security->ScanPolicy does not allow APFS scanning!\n"));
+    ++ErrorCount;
+  }
+
+  if (AsciiAudioDevicePath[0] != '\0') {
```
Comment (vit9696) [inline Utilities/ocvalidate/ValidateUEFI.c:112]: Cannot you reuse the DP validation code?

**Review by vit9696 — COMMENTED** [review]
Marked a few issues.

File: Library/OcMacInfoLib/OcMacInfoLib.c [hunk 5]
```diff
@@ -53,7 +53,8 @@ CONST MAC_INFO_64BIT_COMPAT_ENTRY gMac64BitModels[] = {
 STATIC
 CONST MAC_INFO_INTERNAL_ENTRY *
 LookupInternalEntry (
-  CONST CHAR8    *ProductName
+  CONST CHAR8    *ProductName,
+        BOOLEAN  SetDefaultModel
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Library/OcMacInfoLib/OcMacInfoLib.c]: https://github.com/PMheart/OpenCorePkg/commit/784b6de63b375f62d1f3e323c28ada4946c35858

File: Utilities/ocvalidate/OcValidateLib.c [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+  //
+  gettimeofday (&Time, NULL);
+  //
+  // Return milliseconds.
+  //
+  return Time.tv_sec * 1000LL + Time.tv_usec / 1000LL;
+}
+
+CHAR8 *
+AsciiGetFilenameSuffix (
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/OcValidateLib.c]: Instead, it is done by refactoring `OcAsciiEndsWith`, and `AsciiGetFilenameSuffix` and `AsciiFileNameHasSuffix` are now deprecated.

File: Utilities/ocvalidate/OcValidateLib.c [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+}
+
+BOOLEAN
+AsciiFileSystemPathIsLegal (
+  IN  CONST CHAR8  *Path
+  )
+{
+  UINTN  Index;
+
+  for (Index = 0; Index < AsciiStrLen (Path); ++Index) {
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/OcValidateLib.c]: https://github.com/PMheart/OpenCorePkg/commit/7a3fcbde976d4abd85cd2b2b410b48b68683f302

File: Utilities/ocvalidate/OcValidateLib.c [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    if ((ByteData[Index] & ~ByteMask[Index]) != 0) {
+      return FALSE;
+    }
+  }
+
+  return TRUE;
+}
+
+VOID
+ValidatePatch (
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/OcValidateLib.c:275]: https://github.com/PMheart/OpenCorePkg/commit/4ade41355f5bca94fdbe1dc784e567dc25de5a8e

File: Utilities/ocvalidate/ValidateDeviceProperties.c [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+          DEBUG_WARN,
+          "DeviceProperties->Delete[%u]->Property[%u] contains illegal character!\n",
+          DeviceIndex,
+          PropertyIndex
+          ));
+        ++ErrorCount;
+      }
+    }
+
+    FreePool (DevicePath);
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/ValidateDeviceProperties.c]: https://github.com/PMheart/OpenCorePkg/commit/f9c8de207f6677a33e337256a9c71a3d0b9e75e4

File: Utilities/ocvalidate/ValidateDeviceProperties.c [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+      // Secondly, convert binary back to Unicode device path.
+      //
+      TextualDevicePath = ConvertDevicePathToText (DevicePath, FALSE, FALSE);
+      if (TextualDevicePath != NULL) {
+        //
+        // If the results before and after conversion do not match,
+        // then the original device path is borked.
+        //
+        if (OcStriCmp (UnicodeDevicePath, TextualDevicePath) != 0) {
+          DEBUG ((DEBUG_WARN, "DeviceProperties->Delete[%u]->DevicePath is borked!\n", DeviceIndex));
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/ValidateDeviceProperties.c]: https://github.com/PMheart/OpenCorePkg/commit/f9c8de207f6677a33e337256a9c71a3d0b9e75e4

File: Utilities/ocvalidate/ValidateKernel.c [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+    if (!AsciiFileSystemPathIsLegal (ExecutablePath)) {
+      DEBUG ((DEBUG_WARN, "Kernel->Add[%u]->ExecutablePath contains illegal character!\n", Index));
+      ++ErrorCount;
+      continue;
+    }
+    if (!AsciiFileSystemPathIsLegal (PlistPath)) {
+      DEBUG ((DEBUG_WARN, "Kernel->Add[%u]->PlistPath contains illegal character!\n", Index));
+      ++ErrorCount;
+      continue;
+    } else {
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/ValidateKernel.c]: https://github.com/PMheart/OpenCorePkg/commit/79aa71c11012d7323e0c2468a7c037388f753ccd

File: Utilities/ocvalidate/ValidateKernel.c [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+    }
+    if (!AsciiFileSystemPathIsLegal (BundlePath)) {
+      DEBUG ((DEBUG_WARN, "Kernel->Force[%u]->BundlePath contains illegal character!\n", Index));
+      ++ErrorCount;
+      continue;
+    } else {
+      //
+      // Valid BundlePath must contain .kext suffix.
+      //
+      if (!AsciiFileNameHasSuffix (BundlePath, "kext")) {
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/ValidateKernel.c]: https://github.com/PMheart/OpenCorePkg/commit/79aa71c11012d7323e0c2468a7c037388f753ccd

File: Utilities/ocvalidate/ValidateUEFI.c (hunk 2 above)
Comment (PMheart) [reply Utilities/ocvalidate/ValidateUEFI.c]: https://github.com/PMheart/OpenCorePkg/commit/8b725df764a3ca8481b55133795a998b1bb9cdb4

File: Utilities/ocvalidate/ValidateMisc.c [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+  HibernateMode                  = OC_BLOB_GET (&UserMisc->Boot.HibernateMode);
+  PickerAttributes               = UserMisc->Boot.PickerAttributes;
+  AllowedPickerAttributes        = OC_ATTR_USE_VOLUME_ICON | OC_ATTR_USE_DISK_LABEL_FILE | OC_ATTR_USE_GENERIC_LABEL_IMAGE | OC_ATTR_USE_ALTERNATE_ICONS | OC_ATTR_USE_POINTER_CONTROL;
+  PickerMode                     = OC_BLOB_GET (&UserMisc->Boot.PickerMode);
+  DisplayLevel                   = UserMisc->Debug.DisplayLevel;
+  AllowedDisplayLevel            = DEBUG_WARN | DEBUG_INFO | DEBUG_VERBOSE | DEBUG_ERROR;
+  HaltLevel                      = DisplayLevel;
+  AllowedHaltLevel               = AllowedDisplayLevel;
+  Target                         = UserMisc->Debug.Target;
+  AllowedTarget                  = OC_LOG_ENABLE | OC_LOG_CONSOLE | OC_LOG_DATA_HUB | OC_LOG_SERIAL | OC_LOG_VARIABLE | OC_LOG_NONVOLATILE | OC_LOG_FILE;
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/ValidateMisc.c]: https://github.com/PMheart/OpenCorePkg/commit/e45ac6ae57763fc43cbf54cbbdba502c96f8081c

File: Utilities/ocvalidate/ValidateMisc.c [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+
+  DEBUG ((DEBUG_VERBOSE, "config loaded into Misc checker!\n"));
+
+  ErrorCount                     = 0;
+  UserMisc                       = &Config->Misc;
+  UserUefi                       = &Config->Uefi;
+  ConsoleAttributes              = UserMisc->Boot.ConsoleAttributes;
+  HibernateMode                  = OC_BLOB_GET (&UserMisc->Boot.HibernateMode);
+  PickerAttributes               = UserMisc->Boot.PickerAttributes;
+  AllowedPickerAttributes        = OC_ATTR_USE_VOLUME_ICON | OC_ATTR_USE_DISK_LABEL_FILE | OC_ATTR_USE_GENERIC_LABEL_IMAGE | OC_ATTR_USE_ALTERNATE_ICONS | OC_ATTR_USE_POINTER_CONTROL;
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/ValidateMisc.c]: https://github.com/PMheart/OpenCorePkg/commit/e45ac6ae57763fc43cbf54cbbdba502c96f8081c

File: Utilities/ocvalidate/ValidateMisc.c [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+  AllowedDisplayLevel            = DEBUG_WARN | DEBUG_INFO | DEBUG_VERBOSE | DEBUG_ERROR;
+  HaltLevel                      = DisplayLevel;
+  AllowedHaltLevel               = AllowedDisplayLevel;
+  Target                         = UserMisc->Debug.Target;
+  AllowedTarget                  = OC_LOG_ENABLE | OC_LOG_CONSOLE | OC_LOG_DATA_HUB | OC_LOG_SERIAL | OC_LOG_VARIABLE | OC_LOG_NONVOLATILE | OC_LOG_FILE;
+  BootProtect                    = OC_BLOB_GET (&UserMisc->Security.BootProtect);
+  IsRequestBootVarRoutingEnabled = UserUefi->Quirks.RequestBootVarRouting;
+  AsciiDmgLoading                = OC_BLOB_GET (&UserMisc->Boot.PickerMode);
+  ExposeSensitiveData            = UserMisc->Security.ExposeSensitiveData;
+  AllowedExposeSensitiveData     = OCS_EXPOSE_BOOT_PATH | OCS_EXPOSE_VERSION_VAR | OCS_EXPOSE_VERSION_UI | OCS_EXPOSE_OEM_INFO;
```
[the remark this answers is not in the document]
Comment (PMheart) [reply Utilities/ocvalidate/ValidateMisc.c]: https://github.com/PMheart/OpenCorePkg/commit/e45ac6ae57763fc43cbf54cbbdba502c96f8081c

File: Utilities/ocvalidate/ValidateUEFI.c (hunk 3 above)
Comment (PMheart) [reply Utilities/ocvalidate/ValidateUEFI.c]: https://github.com/PMheart/OpenCorePkg/commit/bebff302b3b15c23f9fcf4f7ba101fa8326cc632

File: Utilities/ocvalidate/ValidateUEFI.c (hunk 4 above)
Comment (PMheart) [reply Utilities/ocvalidate/ValidateUEFI.c]: https://github.com/PMheart/OpenCorePkg/commit/5bbde4c0827dc8c2f18f285b5d0dd7d3fd848832

File: Utilities/ocvalidate/OcValidateLib.c [hunk 16]
```diff
@@ -0,0 +1,422 @@
...
+      DEBUG ((
+        DEBUG_WARN,
+        "%a[%u] has ReplaceMask set but its size is different from Replace (%u vs %u)!\n",
+        PatchSection,
+        PatchIndex,
+        ReplaceMaskSize,
+        ReplaceSize
+        ));
+      ++ErrorCount;
+    } else {
```
Comment (vit9696) [inline Utilities/ocvalidate/OcValidateLib.c:393]: else if maybe?

File: Utilities/ocvalidate/ValidateKernel.c [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+
+  ErrorCount                       = 0;
+  UserKernel                       = &Config->Kernel;
+  UserPlatformInfo                 = &Config->PlatformInfo;
+  IsDisableLinkeditJettisonEnabled = UserKernel->Quirks.DisableLinkeditJettison;
+  IsCustomSMBIOSGuidEnabled        = UserKernel->Quirks.CustomSmbiosGuid;
+  UpdateSMBIOSMode                 = OC_BLOB_GET (&UserPlatformInfo->UpdateSmbiosMode);
+  KernelCache                      = OC_BLOB_GET (&UserKernel->Scheme.KernelCache);
+
+  for (Index = 0; Index < UserKernel->Add.Count; ++Index) {
```
Comment (vit9696) [inline Utilities/ocvalidate/ValidateKernel.c:109]: Maybe add kernel driver and ACPI table duplicate search, just like you did with UEFI drivers?

File: Utilities/ocvalidate/ValidatePlatformInfo.c [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+  ErrorCount         = 0;
+  UserPlatformInfo   = &Config->PlatformInfo;
+  IsAutomaticEnabled = UserPlatformInfo->Automatic;
+  UpdateSMBIOSMode   = OC_BLOB_GET (&UserPlatformInfo->UpdateSmbiosMode);
+  SystemProductName  = OC_BLOB_GET (&UserPlatformInfo->Generic.SystemProductName);
+  SystemMemoryStatus = OC_BLOB_GET (&UserPlatformInfo->Generic.SystemMemoryStatus);
+  AsciiSystemUUID    = OC_BLOB_GET (&UserPlatformInfo->Generic.SystemUuid);
+
+  if (!IsAutomaticEnabled) {
+    DEBUG ((DEBUG_WARN, "PlatformInfo->Automatic is not enabled!\n"));
```
Comment (vit9696) [inline Utilities/ocvalidate/ValidatePlatformInfo.c:50]: This is not a mistake. Please silently quit in this case.

File: Utilities/ocvalidate/OcValidateLib.c (hunk 16 above)
Comment (PMheart) [reply Utilities/ocvalidate/OcValidateLib.c]: https://github.com/PMheart/OpenCorePkg/commit/e390d5b82f664a945e00bcd5ff392589d54f503d

File: Utilities/ocvalidate/ValidatePlatformInfo.c (hunk 18 above)
Comment (PMheart) [reply Utilities/ocvalidate/ValidatePlatformInfo.c]: https://github.com/PMheart/OpenCorePkg/commit/80448d11c0e6ff9860a9301c64435307755be8cf

File: Utilities/ocvalidate/ValidateMisc.c [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+  PickerAttributes               = UserMisc->Boot.PickerAttributes;
+  PickerMode                     = OC_BLOB_GET (&UserMisc->Boot.PickerMode);
+  DisplayLevel                   = UserMisc->Debug.DisplayLevel;
+  AllowedDisplayLevel            = DEBUG_WARN | DEBUG_INFO | DEBUG_VERBOSE | DEBUG_ERROR;
+  HaltLevel                      = DisplayLevel;
+  AllowedHaltLevel               = AllowedDisplayLevel;
+  Target                         = UserMisc->Debug.Target;
+  BootProtect                    = OC_BLOB_GET (&UserMisc->Security.BootProtect);
+  IsRequestBootVarRoutingEnabled = UserUefi->Quirks.RequestBootVarRouting;
+  AsciiDmgLoading                = OC_BLOB_GET (&UserMisc->Boot.PickerMode);
```
Comment (vit9696) [inline Utilities/ocvalidate/ValidateMisc.c:102]: This is a mistake, should be DmgLoading.

Comment (PMheart) [reply Utilities/ocvalidate/ValidateMisc.c]: https://github.com/acidanthera/OpenCorePkg/pull/169/commits/08bdbce451c29b362124f2a71603740a86beb343

**Pull request merged**